openmessaging-api/src/main/java/io/openmessaging/api/Client.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2017 OpenMessaging
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.openmessaging.api;
+
+import io.openmessaging.api.consumer.Consumer;
+import io.openmessaging.api.consumer.ConsumerConfiguration;
+import io.openmessaging.api.producer.Producer;
+import io.openmessaging.api.producer.ProducerConfiguration;
+import io.openmessaging.api.reader.Reader;
+import io.openmessaging.api.reader.ReaderConfiguration;
+import io.openmessaging.exception.OMSException;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An OMS client that talks to OMS messaging services.
+ */
+public interface Client extends AutoCloseable {
+
+    /**
+     * Create a producer with given {@code ProducerConfiguration} for publishing on a specific topic
+     *
+     * @param topic
+     *            The name of the topic where to produce
+     * @param conf
+     *            The {@code ProducerConfiguration} object
+     * @return The producer object
+     * @throws OMSException
+     *             if it was not possible to create the producer
+     * @throws InterruptedException
+     */
+    Producer createProducer(String topic, ProducerConfiguration conf) throws OMSException;
+
+    /**
+     * Asynchronously create a producer with given {@code ProducerConfiguration} for publishing on a specific topic
+     *
+     * @param topic
+     *            The name of the topic where to produce
+     * @param conf
+     *            The {@code ProducerConfiguration} object
+     * @return Future of the asynchronously created producer object
+     */
+    CompletableFuture<Producer> createProducerAsync(String topic, ProducerConfiguration conf);
```

Comment:
We have defined the interface `MessagingAccessPoint` to create, update and list the producers/consumers, and the `MessagingAccessPoint` instance is generated from OMS driver URL.

And, IMO, there is no need to provide an async method to create clients, the clients will be created at preparation stage in most cases.

Reply:
@zhouxinyu yes `Client` is what you called `MessagingAccessPoint`. but why not keep things simpler, call it `Client`. 

The async method are required when you are using a client to create multiple producers on-the-fly. you don't want to be blocked on creating a producer and consumer.

Reply:
The same reason, I prefer fine-grained definiation :-)

Reply:
@vongosling @zhouxinyu please tell me what is the difference between MessagingAccessPoint and Client here. It is the instance that application used for creating producer, consumer.